### PR TITLE
i16x16

### LINKED ISSUE #59: u16x16
This one fits into 256bits, so it can be implemented via AVX.

### PR DESCRIPTION
**Pull request by gilescope**
Doesn't fix #59 but does add i16x16.
(For the test data I've used 32x8's test data twice.)

### CODE REVIEW

File: src/i16x16_.rs [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+  #[inline]
+  #[must_use]
+  fn cmp_lt(self, rhs: Self) -> Self::Output {
+    pick! {
+      if #[cfg(target_feature="avx2")] {
+        //TODO: what about == ?
+        Self { avx2: !cmp_gt_mask_i16_m256i(self.avx2, rhs.avx2) }
+      } else if #[cfg(target_feature="sse2")] {
+        Self { sse0: cmp_lt_mask_i16_m128i(self.sse0, rhs.sse0),
+          sse1: cmp_lt_mask_i16_m128i(self.sse1, rhs.sse1) }
```
Comment (gilescope) [inline src/i16x16_.rs:405]: am suspicious of this. !gt is not the same as lt. I've done the same as others but wondering if this is a bug?

File: src/i16x16_.rs [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+
+impl i16x16 {
+  #[inline]
+  #[must_use]
+  pub fn blend(self, t: Self, f: Self) -> Self {
+    pick! {
+      if #[cfg(target_feature="avx2")] {
+        Self { avx2: blend_varying_i8_m256i(f.avx2, t.avx2, self.avx2) }
+      } else if #[cfg(target_feature="sse4.1")] {
+        Self { sse: blend_varying_i8_m128i(f.sse, t.sse, self.sse) }
```
Comment (gilescope) [inline src/i16x16_.rs:438]: Is this right that blend should be i8?

File: src/i16x16_.rs (hunk 1 above)
Comment (Lokathor) [reply src/i16x16_.rs:405]: yeah !gt isn't the same as lt, because two i16 can be equal

File: src/i16x16_.rs (hunk 2 above)
Comment (Lokathor) [reply src/i16x16_.rs:438]: an i8 blend should be fine if the mask is an i16 mask. each two i8 lanes of the mask will be identical, so they'll use the identical source for the output, so it ends up working like an i16 blend.

File: src/i16x16_.rs (hunk 1 above)
Comment (gilescope) [reply src/i16x16_.rs:404]: Excellent, I've fixed the i32x8 code that I was basing it from. Was a lucky chance that I happened to copy that one as that seemed the only one that wasn't correct.

**Comment by gilescope** [comment]
Does that look ok @Lokathor or would you like u16x16 also? (If you do that's fine though I'd be tempted to put that in a separate PR)

**Comment by Lokathor** [comment]
skipping the u16x16 stuff is fine.

I will try to have a look here later today, but in the meantime you probably are missing a few cfg things since the i586 had trouble in Ci.

File: src/i16x16_.rs [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+              self.arr[6] >> u,
+              self.arr[7] >> u,
+              self.arr[8] << u,
+              self.arr[9] << u,
+              self.arr[10] << u,
+              self.arr[11] << u,
+              self.arr[12] << u,
+              self.arr[13] << u,
+              self.arr[14] << u,
+              self.arr[15] << u,
```
Comment (Lokathor) [inline src/i16x16_.rs:314]: 8 to 15 are still shifting left

**Review by Lokathor — CHANGES_REQUESTED** [review]

File: src/i16x16_.rs (hunk 3 above)
Comment (gilescope) [reply src/i16x16_.rs:314]: ok will add more tests to pick that up.

Comment (gilescope) [reply src/i16x16_.rs:314]: (good spot)

Comment (gilescope) [reply src/i16x16_.rs]: Ah I see that the test was there and it was caught by CI. nice.

**Review by Lokathor — APPROVED** [review]

**Pull request merged**